client/src/main/java/org/eclipse/hono/client/impl/ProtocolAdapterCommandConsumerFactoryImpl.java

```diff
@@ -208,7 +223,18 @@ protected void onDisconnect() {
                             deviceId, thr);
                     return Future.failedFuture(thr);
                 })
-                .mapEmpty();
+                .compose(removed -> {
+                    final boolean entryNotExpired = lifespan.isNegative() || Instant.now().isBefore(lifespanStart.plus(lifespan));
+                    if (!removed && entryNotExpired) {
+                        // entry wasn't actually removed and entry hasn't expired (yet)
```

Comment:
this covers adapters which use stateful protocols like AMQP and MQTT where the TTD is -1, right?
again, IMHO it would be helpful to describe the circumstances under which this happens here in the comment in some more detail

Reply:
> this covers adapters which use stateful protocols like AMQP and MQTT where the TTD is -1, right?

Most importantly, but not exclusively. There could also be 2 overlapping HTTP adapter TTD requests, the above "if" block getting reached when calling `removeCommandConsumer` on the 1st request before the 2nd ttd period has elapsed. It's not that important throwing an exception in that case though, since the HTTP adapter won't send a "disconnectedTtd" event. But could be informative in the tracing span anyway.